swayidle/main.c

```diff
@@ -0,0 +1,481 @@
+#define _XOPEN_SOURCE 500
+#include <getopt.h>
+#include <signal.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+#include <wayland-client-protocol.h>
+#include <wayland-client.h>
+#include <wayland-util.h>
+#include <wlr/config.h>
+#include <wlr/util/log.h>
+#include <wlr/types/wlr_output_layout.h>
+#include <wlr/types/wlr_output.h>
+#include "idle-client-protocol.h"
+#include "ipc-client.h"
+#include <json-c/json.h>
+#include "config.h"
+#include "list.h"
+#ifdef SWAY_IDLE_HAS_SYSTEMD
+#include <systemd/sd-bus.h>
+#include <systemd/sd-login.h>
+#elif defined(SWAY_IDLE_HAS_ELOGIND)
+#include <elogind/sd-bus.h>
+#include <elogind/sd-login.h>
+#endif
+
+typedef void (*timer_callback_func)(void *data);
+
+static struct org_kde_kwin_idle *idle_manager = NULL;
+static struct wl_seat *seat = NULL;
+static int ipc_socketfd = -1;
+char *socket_path = NULL;
+bool debug = false;
+bool setup_inhibit = true;
+int inhibit_cnt = 0;
+
+struct swayidle_state {
+	struct wl_display *display;
+	struct org_kde_kwin_idle_timeout *idle_timer;
+	struct org_kde_kwin_idle_timeout *lock_timer;
+	struct wlr_output_layout *layout;
+	struct wl_event_loop *event_loop;
+	list_t *timeout_cmds;
+} state;
+
+struct swayidle_cmd {
+	timer_callback_func callback;
+	char *param;
+};
+
+struct swayidle_timeout_cmd {
+	uint32_t timeout;
+	struct swayidle_cmd *idle_cmd;
+	struct swayidle_cmd *resume_cmd;
+};
+
+static void cmd_exec(void *data) {
+	if (data == NULL)
+		return;
+	char *param = (char *) data;
+	wlr_log(L_INFO,"Cmd exec %s", param);
+	int pid = fork();
+	if (pid == 0) {
+		char *const cmd[] = { "sh", "-c", param, NULL, };
+		execvp(cmd[0], cmd);
+		exit(1);
+	}
+	wlr_log(L_DEBUG, "Spawned process %d", pid);
+}
+
+static void cmd_command(void *data) {
+	if (data == NULL)
+		return;
+	char *param = (char *) data;
+	wlr_log(L_DEBUG,"Cmd command %s", param);
+	uint32_t size = strlen(param);
+	ipc_single_command(ipc_socketfd, IPC_COMMAND, param, &size);
+}
+
+#if defined(SWAY_IDLE_HAS_SYSTEMD) || defined(SWAY_IDLE_HAS_ELOGIND)
+static int lock_fd = -1;
+static int ongoing_fd = -1;
+
+static int release_lock(void *data) {
+	wlr_log(L_INFO, "Releasing sleep lock %d", ongoing_fd);
+	if (ongoing_fd>=0) {
+		close(ongoing_fd);
+	}
+	ongoing_fd = -1;
+	return 0;
+}
+
+bool have_lock() {
+	uint32_t len = 0;
+	char *res = ipc_single_command(ipc_socketfd, IPC_IS_LOCKSCREEN_ACTIVE, NULL, &len);
+	json_object *is_active_json = json_tokener_parse(res);
+	bool locked = json_object_get_boolean(is_active_json);
+	wlr_log(L_DEBUG,"Got lock? %d", locked);
+	return locked;
+}
```

Comment:
What do you need to know this for?

Reply:
That is to know when the lockscreen have been raised, so I can release the sleep lock.

Reply:
I suppose I'm not clear on what the sleep lock is supposed to do.

Reply:
The sleep lock prevents the laptop from going to sleep until it is release. We use this to ensure that the lockscreen is up before the system is going to sleep. If we do not release the lock, the system will have a delay on about 5 seconds before going to sleep. 
If we do not use a sleep lock, the system may go to sleep before the lock is up. This will cause any person that opens the laptop to see the content on the screen on resume, before the lockscreen is up.

Reply:
I see. Rather than an IPC command can't we just assume that if we ran swaylock the lock screen is up?